File: backend/app/services/service_status.py
"""
Service Status Collector.

Aggregates status information from all background services for the admin
debugging dashboard. Provides unified access to service health, dependencies,
and application metrics.
"""

from __future__ import annotations

import asyncio
import logging
import os
import shutil
import time
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional, Tuple, Callable, Any

import psutil

from app.core.config import settings
from app.schemas.service_status import (
    ServiceStateEnum,
    ServiceStatusResponse,
    DependencyStatusResponse,
    ApplicationMetricsResponse,
    AdminDebugResponse,
    DbPoolStatus,
    CacheStatsResponse,
    ServiceRestartResponse,
    ServiceStopResponse,
    ServiceStartResponse,
)

logger = logging.getLogger(__name__)

# Server start time for uptime calculation
_SERVER_START_TIME: Optional[float] = None


def set_server_start_time() -> None:
    """Set the server start time. Call this during app startup."""
    global _SERVER_START_TIME
    _SERVER_START_TIME = time.time()


def get_server_uptime() -> float:
    """Get server uptime in seconds."""
    if _SERVER_START_TIME is None:
        return 0.0
    return time.time() - _SERVER_START_TIME


# Service registry - maps service name to status getter and restart functions
_service_registry: Dict[str, Dict[str, Any]] = {}


def register_service(
    name: str,
    display_name: str,
    get_status_fn: Callable[[], Dict[str, Any]],
    restart_fn: Optional[Callable[[], Any]] = None,
    stop_fn: Optional[Callable[[], Any]] = None,
    start_fn: Optional[Callable[[], Any]] = None,
    config_enabled_fn: Optional[Callable[[], bool]] = None,
) -> None:
    """
    Register a service for status monitoring.

    Args:
        name: Service identifier (e.g., 'telemetry_monitor')
        display_name: Human-readable name (e.g., 'Telemetry Monitor')
        get_status_fn: Function that returns service status dict
        restart_fn: Optional function to restart the service
        stop_fn: Optional function to stop the service
        start_fn: Optional function to start the service
        config_enabled_fn: Optional function to check if service is enabled in config
    """
    _service_registry[name] = {
        "display_name": display_name,
        "get_status": get_status_fn,
        "restart": restart_fn,
        "stop": stop_fn,
        "start": start_fn,
        "config_enabled": config_enabled_fn,
    }
    logger.debug(f"Registered service: {name}")


def unregister_service(name: str) -> None:
    """Unregister a service from monitoring."""
    if name in _service_registry:
        del _service_registry[name]
        logger.debug(f"Unregistered service: {name}")


class ServiceStatusCollector:
    """Aggregates status from all registered background services."""

    def __init__(self):
        self._db_engine = None
        self._service_locks: Dict[str, asyncio.Lock] = {}

    def set_db_engine(self, engine) -> None:
        """Set the database engine for connection pool monitoring."""
        self._db_engine = engine

    def _get_service_lock(self, service_name: str) -> asyncio.Lock:
        """
        Get or create a lock for a specific service.

        Ensures only one control operation (start/stop/restart) can run
        at a time for each service.
        """
        if service_name not in self._service_locks:
            self._service_locks[service_name] = asyncio.Lock()
        return self._service_locks[service_name]

    def get_all_services(self) -> List[ServiceStatusResponse]:
        """
        Get status for all registered services.

        Returns:
            List of ServiceStatusResponse objects
        """
        services = []

        for name, registry in _service_registry.items():
            try:
                status_fn = registry.get("get_status")
                if status_fn is None:
                    continue

                # Call the status function
                status_data = status_fn()

                # Determine state
                state = ServiceStateEnum.RUNNING
                if status_data.get("is_running") is False:
                    state = ServiceStateEnum.STOPPED
                elif status_data.get("has_error"):
                    state = ServiceStateEnum.ERROR

                # Check if service is enabled in config
                config_enabled = True
                config_enabled_fn = registry.get("config_enabled")
                if config_enabled_fn:
                    config_enabled = config_enabled_fn()

                if not config_enabled:
                    state = ServiceStateEnum.DISABLED

                # Build response
                service_response = ServiceStatusResponse(
                    name=name,
                    display_name=registry["display_name"],
                    state=state,
                    started_at=status_data.get("started_at"),
                    uptime_seconds=status_data.get("uptime_seconds"),
                    sample_count=status_data.get("sample_count"),
                    error_count=status_data.get("error_count", 0),
                    last_error=status_data.get("last_error"),
                    last_error_at=status_data.get("last_error_at"),
                    config_enabled=config_enabled,
                    interval_seconds=status_data.get("interval_seconds"),
                    restartable=registry.get("restart") is not None or (
                        registry.get("stop") is not None and registry.get("start") is not None
                    ),
                )
                services.append(service_response)

            except Exception as e:
                logger.error(f"Error getting status for service {name}: {e}")
                # Return error state for this service
                services.append(ServiceStatusResponse(
                    name=name,
                    display_name=registry.get("display_name", name),
                    state=ServiceStateEnum.ERROR,
                    error_count=1,
                    last_error=str(e),
                    last_error_at=datetime.utcnow(),
                    config_enabled=True,
                    restartable=False,
                ))

        return services

    def get_service(self, name: str) -> Optional[ServiceStatusResponse]:
        """
        Get status for a specific service.

        Args:
            name: Service identifier

        Returns:
            ServiceStatusResponse or None if service not found
        """
        if name not in _service_registry:
            return None

        services = self.get_all_services()
        for service in services:
            if service.name == name:
                return service
        return None

    async def restart_service(self, name: str, force: bool = False) -> ServiceRestartResponse:
        """
        Restart a specific service.

        Args:
            name: Service identifier
            force: Force restart even if service appears healthy

        Returns:
            ServiceRestartResponse with result
        """
        if name not in _service_registry:
            return ServiceRestartResponse(
                success=False,
                service_name=name,
                previous_state=ServiceStateEnum.STOPPED,
                current_state=ServiceStateEnum.STOPPED,
                message=f"Service '{name}' not found",
            )

        # Acquire lock to prevent concurrent operations on this service
        async with self._get_service_lock(name):
            registry = _service_registry[name]

            # Get current state
            current_service = self.get_service(name)
            previous_state = current_service.state if current_service else ServiceStateEnum.STOPPED

            # Check if restartable
            restart_fn = registry.get("restart")
            stop_fn = registry.get("stop")
            start_fn = registry.get("start")

            if restart_fn is None and (stop_fn is None or start_fn is None):
                return ServiceRestartResponse(
                    success=False,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=previous_state,
                    message=f"Service '{name}' is not restartable",
                )

            try:
                # Use dedicated restart function if available
                if restart_fn:
                    result = restart_fn()
                    if asyncio.iscoroutine(result):
                        await result
                else:
                    # Stop then start
                    if stop_fn:
                        result = stop_fn()
                        if asyncio.iscoroutine(result):
                            await result

                    # Small delay between stop and start
                    await asyncio.sleep(0.5)

                    if start_fn:
                        result = start_fn()
                        if asyncio.iscoroutine(result):
                            await result

                # Get new state
                await asyncio.sleep(0.5)  # Allow service to stabilize
                new_service = self.get_service(name)
                new_state = new_service.state if new_service else ServiceStateEnum.STOPPED

                logger.info(f"Service '{name}' restarted: {previous_state} -> {new_state}")

                return ServiceRestartResponse(
                    success=True,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=new_state,
                    message="Service restarted successfully",
                )

            except Exception as e:
                logger.error(f"Error restarting service {name}: {e}")
                new_service = self.get_service(name)
                new_state = new_service.state if new_service else ServiceStateEnum.ERROR

                return ServiceRestartResponse(
                    success=False,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=new_state,
                    message=str(e),
                )

    async def stop_service(self, name: str, force: bool = False) -> ServiceStopResponse:
        """
        Stop a specific service.

        Args:
            name: Service identifier
            force: Force stop even if service appears healthy

        Returns:
            ServiceStopResponse with result
        """
        if name not in _service_registry:
            return ServiceStopResponse(
                success=False,
                service_name=name,
                previous_state=ServiceStateEnum.STOPPED,
                current_state=ServiceStateEnum.STOPPED,
                message=f"Service '{name}' not found",
            )

        # Acquire lock to prevent concurrent operations on this service
        async with self._get_service_lock(name):
            registry = _service_registry[name]

            # Get current state
            current_service = self.get_service(name)
            previous_state = current_service.state if current_service else ServiceStateEnum.STOPPED

            # Check if stoppable
            stop_fn = registry.get("stop")
            if stop_fn is None:
                return ServiceStopResponse(
                    success=False,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=previous_state,
                    message=f"Service '{name}' does not support stop operation",
                )

            try:
                # Call stop function
                result = stop_fn()
                if asyncio.iscoroutine(result):
                    await result

                # Get new state
                await asyncio.sleep(0.5)  # Allow service to stabilize
                new_service = self.get_service(name)
                new_state = new_service.state if new_service else ServiceStateEnum.STOPPED

                logger.info(f"Service '{name}' stopped: {previous_state} -> {new_state}")

                return ServiceStopResponse(
                    success=True,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=new_state,
                    message="Service stopped successfully",
                )

            except Exception as e:
                logger.error(f"Error stopping service {name}: {e}")
                new_service = self.get_service(name)
                new_state = new_service.state if new_service else ServiceStateEnum.ERROR

                return ServiceStopResponse(
                    success=False,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=new_state,
                    message=str(e),
                )

    async def start_service(self, name: str, force: bool = False) -> ServiceStartResponse:
        """
        Start a specific service.

        Args:
            name: Service identifier
            force: Force start even if service appears running

        Returns:
            ServiceStartResponse with result
        """
        if name not in _service_registry:
            return ServiceStartResponse(
                success=False,
                service_name=name,
                previous_state=ServiceStateEnum.STOPPED,
                current_state=ServiceStateEnum.STOPPED,
                message=f"Service '{name}' not found",
            )

        # Acquire lock to prevent concurrent operations on this service
        async with self._get_service_lock(name):
            registry = _service_registry[name]

            # Get current state
            current_service = self.get_service(name)
            previous_state = current_service.state if current_service else ServiceStateEnum.STOPPED

            # Check if startable
            start_fn = registry.get("start")
            if start_fn is None:
                return ServiceStartResponse(
                    success=False,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=previous_state,
                    message=f"Service '{name}' does not support start operation",
                )

            try:
                # Call start function
                result = start_fn()
                if asyncio.iscoroutine(result):
                    await result

                # Get new state
                await asyncio.sleep(0.5)  # Allow service to stabilize
                new_service = self.get_service(name)
                new_state = new_service.state if new_service else ServiceStateEnum.STOPPED

                logger.info(f"Service '{name}' started: {previous_state} -> {new_state}")

                return ServiceStartResponse(
                    success=True,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=new_state,
                    message="Service started successfully",
                )

            except Exception as e:
                logger.error(f"Error starting service {name}: {e}")
                new_service = self.get_service(name)
                new_state = new_service.state if new_service else ServiceStateEnum.ERROR

                return ServiceStartResponse(
                    success=False,
                    service_name=name,
                    previous_state=previous_state,
                    current_state=new_state,
                    message=str(e),
                )

    def get_dependencies(self) -> List[DependencyStatusResponse]:
        """
        Check availability of system dependencies.

        Returns:
            List of DependencyStatusResponse objects
        """
        dependencies = []

        # smartctl - SMART disk health
        smartctl_path = shutil.which("smartctl")
        dependencies.append(DependencyStatusResponse(
            name="smartctl",
            available=smartctl_path is not None,
            path=smartctl_path,
            version=self._get_tool_version(smartctl_path, ["--version"]) if smartctl_path else None,
            required_for=["SMART Disk Health", "Disk Monitoring"],
        ))

        # mdadm - RAID management
        mdadm_path = shutil.which("mdadm")
        dependencies.append(DependencyStatusResponse(
            name="mdadm",
            available=mdadm_path is not None,
            path=mdadm_path,
            version=self._get_tool_version(mdadm_path, ["--version"]) if mdadm_path else None,
            required_for=["RAID Management"],
        ))

        # wg (WireGuard) - VPN
        wg_path = shutil.which("wg")
        dependencies.append(DependencyStatusResponse(
            name="wg",
            available=wg_path is not None,
            path=wg_path,
            version=self._get_tool_version(wg_path, ["--version"]) if wg_path else None,
            required_for=["VPN Configuration"],
        ))

        # ipmitool - Fan control (optional)
        ipmitool_path = shutil.which("ipmitool")
        dependencies.append(DependencyStatusResponse(
            name="ipmitool",
            available=ipmitool_path is not None,
            path=ipmitool_path,
            version=self._get_tool_version(ipmitool_path, ["-V"]) if ipmitool_path else None,
            required_for=["IPMI Fan Control (optional)"],
        ))

        # hwmon/sysfs - Temperature sensors
        hwmon_available = Path("/sys/class/hwmon").exists()
        dependencies.append(DependencyStatusResponse(
            name="hwmon",
            available=hwmon_available,
            path="/sys/class/hwmon" if hwmon_available else None,
            version=None,
            required_for=["Temperature Sensors", "Fan Control"],
        ))

        # cpufreq - CPU frequency scaling
        cpufreq_available = Path("/sys/devices/system/cpu/cpu0/cpufreq").exists()
        dependencies.append(DependencyStatusResponse(
            name="cpufreq",
            available=cpufreq_available,
            path="/sys/devices/system/cpu/cpu0/cpufreq" if cpufreq_available else None,
            version=None,
            required_for=["CPU Power Management"],
        ))

        return dependencies

    def _get_tool_version(self, path: str, args: List[str]) -> Optional[str]:
        """Get version string for a tool."""
        if not path:
            return None

        import subprocess
        try:
            result = subprocess.run(
                [path] + args,
                capture_output=True,
                text=True,
                timeout=5
            )
            output = result.stdout or result.stderr
            # Extract first line which usually contains version
            if output:
                first_line = output.strip().split("\n")[0]
                return first_line[:100]  # Limit length
        except Exception:
            pass
        return None

    def get_app_metrics(self) -> ApplicationMetricsResponse:
        """
        Get application-level metrics.

        Returns:
            ApplicationMetricsResponse with current metrics
        """
        # Import error counter
        from app.middleware.error_counter import get_error_counts

        # Get error counts
        error_4xx, error_5xx = get_error_counts()

        # Get active asyncio tasks
        try:
            active_tasks = len(asyncio.all_tasks())
        except RuntimeError:
            active_tasks = 0

        # Get process memory
        try:
            process = psutil.Process(os.getpid())
            memory_info = process.memory_info()
            memory_bytes = memory_info.rss
            memory_percent = process.memory_percent()
        except Exception:
            memory_bytes = 0
            memory_percent = 0.0

        # Get DB pool status
        db_pool_status = None
        if self._db_engine is not None:
            try:
                pool = self._db_engine.pool
                db_pool_status = DbPoolStatus(
                    pool_size=pool.size(),
                    checked_in=pool.checkedin(),
                    checked_out=pool.checkedout(),
                    overflow=pool.overflow(),
                )
            except Exception:
                pass

        # Collect cache stats from services that have them
        cache_stats = []
        # SMART cache
        try:
            from app.services import smart
            if hasattr(smart, 'get_cache_stats'):
                stats = smart.get_cache_stats()
                if stats:
                    cache_stats.append(CacheStatsResponse(
                        name="smart_cache",
                        hits=stats.get("hits", 0),
                        misses=stats.get("misses", 0),
                        size=stats.get("size", 0),
                        max_size=stats.get("max_size"),
                    ))
        except Exception:
            pass

        # RAID cache
        try:
            from app.services import raid
            if hasattr(raid, 'get_cache_stats'):
                stats = raid.get_cache_stats()
                if stats:
                    cache_stats.append(CacheStatsResponse(
                        name="raid_cache",
                        hits=stats.get("hits", 0),
                        misses=stats.get("misses", 0),
                        size=stats.get("size", 0),
                        max_size=stats.get("max_size"),
                    ))
        except Exception:
            pass

        return ApplicationMetricsResponse(
            server_uptime_seconds=get_server_uptime(),
            error_count_4xx=error_4xx,
            error_count_5xx=error_5xx,
            active_tasks=active_tasks,
            memory_bytes=memory_bytes,
            memory_percent=round(memory_percent, 2),
            db_pool_status=db_pool_status,
            cache_stats=cache_stats,
        )

    def get_debug_snapshot(self) -> AdminDebugResponse:
        """
        Get combined debug snapshot for admin dashboard.

        Returns:
            AdminDebugResponse with all status information
        """
        return AdminDebugResponse(
            timestamp=datetime.utcnow(),
            services=self.get_all_services(),
            dependencies=self.get_dependencies(),
            metrics=self.get_app_metrics(),
        )


# Singleton instance
_collector: Optional[ServiceStatusCollector] = None


def get_service_status_collector() -> ServiceStatusCollector:
    """Get the singleton ServiceStatusCollector instance."""
    global _collector
    if _collector is None:
        _collector = ServiceStatusCollector()
    return _collector

File: client/src/api/service-status.ts
/**
 * Service status API client for admin debugging dashboard
 */
import { apiClient } from '../lib/api';
import { formatBytes as sharedFormatBytes } from '../lib/formatters';

// Enums
export const ServiceState = {
  RUNNING: 'running',
  STOPPED: 'stopped',
  ERROR: 'error',
  DISABLED: 'disabled'
} as const;
export type ServiceState = typeof ServiceState[keyof typeof ServiceState];

// Types
export interface ServiceStatus {
  name: string;
  display_name: string;
  state: ServiceState;
  started_at: string | null;
  uptime_seconds: number | null;
  sample_count: number | null;
  error_count: number;
  last_error: string | null;
  last_error_at: string | null;
  config_enabled: boolean;
  interval_seconds: number | null;
  restartable: boolean;
}

export interface DependencyStatus {
  name: string;
  available: boolean;
  path: string | null;
  version: string | null;
  required_for: string[];
}

export interface DbPoolStatus {
  pool_size: number;
  checked_in: number;
  checked_out: number;
  overflow: number;
}

export interface CacheStats {
  name: string;
  hits: number;
  misses: number;
  size: number;
  max_size: number | null;
}

export interface ApplicationMetrics {
  server_uptime_seconds: number;
  error_count_4xx: number;
  error_count_5xx: number;
  active_tasks: number;
  memory_bytes: number;
  memory_percent: number;
  db_pool_status: DbPoolStatus | null;
  cache_stats: CacheStats[];
}

export interface AdminDebugSnapshot {
  timestamp: string;
  services: ServiceStatus[];
  dependencies: DependencyStatus[];
  metrics: ApplicationMetrics;
}

export interface ServiceRestartRequest {
  force?: boolean;
}

export interface ServiceRestartResponse {
  success: boolean;
  service_name: string;
  previous_state: ServiceState;
  current_state: ServiceState;
  message: string | null;
}

export interface ServiceStopRequest {
  force?: boolean;
}

export interface ServiceStopResponse {
  success: boolean;
  service_name: string;
  previous_state: ServiceState;
  current_state: ServiceState;
  message: string | null;
}

export interface ServiceStartRequest {
  force?: boolean;
}

export interface ServiceStartResponse {
  success: boolean;
  service_name: string;
  previous_state: ServiceState;
  current_state: ServiceState;
  message: string | null;
}

// API Functions

/**
 * Get status for all registered services (admin only)
 */
export async function getAllServices(): Promise<ServiceStatus[]> {
  const response = await apiClient.get<ServiceStatus[]>('/api/admin/services');
  return response.data;
}

/**
 * Get status for a specific service (admin only)
 */
export async function getService(serviceName: string): Promise<ServiceStatus> {
  const response = await apiClient.get<ServiceStatus>(`/api/admin/services/${serviceName}`);
  return response.data;
}

/**
 * Restart a specific service (admin only)
 */
export async function restartService(
  serviceName: string,
  force: boolean = false
): Promise<ServiceRestartResponse> {
  const request: ServiceRestartRequest = { force };
  const response = await apiClient.post<ServiceRestartResponse>(
    `/api/admin/services/${serviceName}/restart`,
    request
  );
  return response.data;
}

/**
 * Stop a specific service (admin only)
 */
export async function stopService(
  serviceName: string,
  force: boolean = false
): Promise<ServiceStopResponse> {
  const request: ServiceStopRequest = { force };
  const response = await apiClient.post<ServiceStopResponse>(
    `/api/admin/services/${serviceName}/stop`,
    request
  );
  return response.data;
}

/**
 * Start a specific service (admin only)
 */
export async function startService(
  serviceName: string,
  force: boolean = false
): Promise<ServiceStartResponse> {
  const request: ServiceStartRequest = { force };
  const response = await apiClient.post<ServiceStartResponse>(
    `/api/admin/services/${serviceName}/start`,
    request
  );
  return response.data;
}

/**
 * Get system dependency availability (admin only)
 */
export async function getDependencies(): Promise<DependencyStatus[]> {
  const response = await apiClient.get<DependencyStatus[]>('/api/admin/dependencies');
  return response.data;
}

/**
 * Get application-level metrics (admin only)
 */
export async function getApplicationMetrics(): Promise<ApplicationMetrics> {
  const response = await apiClient.get<ApplicationMetrics>('/api/admin/metrics');
  return response.data;
}

/**
 * Get combined debug snapshot (admin only)
 */
export async function getDebugSnapshot(): Promise<AdminDebugSnapshot> {
  const response = await apiClient.get<AdminDebugSnapshot>('/api/admin/debug');
  return response.data;
}

// Helper functions

/**
 * Format uptime to human-readable string
 */
export function formatUptime(seconds: number | null): string {
  if (seconds === null || seconds === undefined) return '-';

  const days = Math.floor(seconds / 86400);
  const hours = Math.floor((seconds % 86400) / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const secs = Math.floor(seconds % 60);

  if (days > 0) {
    return `${days}d ${hours}h ${minutes}m`;
  } else if (hours > 0) {
    return `${hours}h ${minutes}m ${secs}s`;
  } else if (minutes > 0) {
    return `${minutes}m ${secs}s`;
  }
  return `${secs}s`;
}

/**
 * Format bytes to human-readable string (re-export from shared formatters)
 */
export const formatBytes = sharedFormatBytes;

/**
 * Get state color for UI display
 */
export function getStateColor(state: ServiceState): string {
  switch (state) {
    case ServiceState.RUNNING:
      return 'text-green-500';
    case ServiceState.STOPPED:
      return 'text-gray-500';
    case ServiceState.ERROR:
      return 'text-red-500';
    case ServiceState.DISABLED:
      return 'text-yellow-500';
    default:
      return 'text-gray-500';
  }
}

/**
 * Get state background color for badges
 */
export function getStateBgColor(state: ServiceState): string {
  switch (state) {
    case ServiceState.RUNNING:
      return 'bg-green-100 text-green-800 dark:bg-green-900 dark:text-green-200';
    case ServiceState.STOPPED:
      return 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-200';
    case ServiceState.ERROR:
      return 'bg-red-100 text-red-800 dark:bg-red-900 dark:text-red-200';
    case ServiceState.DISABLED:
      return 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900 dark:text-yellow-200';
    default:
      return 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-200';
  }
}
